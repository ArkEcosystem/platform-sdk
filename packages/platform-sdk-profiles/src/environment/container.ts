export class Container {
	readonly #bindings: Map<string, any> = new Map();

	public get<T>(key: string): T {
		if (this.has(key)) {
			return this.#bindings.get(key);
		}

		throw new Error(`No matching bindings found for [${key}].`);
	}

	public bind(key: string, value: unknown): void {
		if (this.has(key)) {
			throw new Error(`Binding with name [${key}] already exists. Call [rebind] to replace it.`);
		}

		this.#bindings.set(key, value);
	}

	public rebind(key: string, value: unknown): void {
		this.unbind(key);

		this.bind(key, value);
	}

	public has(key: string): boolean {
		return this.#bindings.has(key);
	}

	public unbind(key: string): void {
		if (this.has(key)) {
			this.#bindings.delete(key);

			return;
		}

		throw new Error(`No matching bindings found for [${key}].`);
	}

	public flush(): void {
		for (const key of this.#bindings.keys()) {
			this.unbind(key);
		}
	}
}

export const container = new Container();
