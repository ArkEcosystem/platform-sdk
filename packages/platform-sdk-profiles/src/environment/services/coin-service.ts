import { Coins } from "@arkecosystem/platform-sdk";

import { DataRepository } from "../../repositories/data-repository";
import { container } from "../container";
import { Identifiers } from "../container.models";

export class CoinService {
	readonly #dataRepository: DataRepository = new DataRepository();

	public all(): Record<string, Coins.Coin> {
		return this.#dataRepository.all() as Record<string, Coins.Coin>;
	}

	public values(): Coins.Coin[] {
		return this.#dataRepository.values();
	}

	public entries(): [string, string[]][] {
		const result: Record<string, string[]> = {};

		for (const [coin, networks] of Object.entries(this.all())) {
			result[coin] = [];

			for (const [network, children] of Object.entries(networks)) {
				if (children !== undefined) {
					for (const child of Object.keys(children)) {
						result[coin].push(`${network}.${child}`);
					}
				} else {
					result[coin].push(network);
				}
			}
		}

		return Object.entries(result);
	}

	public get(coin: string, network: string): Coins.Coin {
		const instance: Coins.Coin | undefined = this.#dataRepository.get<Coins.Coin>(`${coin}.${network}`);

		if (instance === undefined) {
			throw new Error(`An instance for [${coin}.${network}] does not exist.`);
		}

		return instance;
	}

	public async push(coin: string, network: string, options: object = {}, useForce = false): Promise<Coins.Coin> {
		if (!useForce && this.has(coin, network)) {
			return this.get(coin, network);
		}

		this.#dataRepository.set(
			`${coin}.${network}`,
			await Coins.CoinFactory.make(container.get<Coins.CoinSpec>(Identifiers.Coins)[coin.toUpperCase()], {
				network,
				httpClient: container.get(Identifiers.HttpClient),
				...options,
			}),
		);

		return this.get(coin, network);
	}

	public has(coin: string, network: string): boolean {
		return this.#dataRepository.has(`${coin}.${network}`);
	}
}
