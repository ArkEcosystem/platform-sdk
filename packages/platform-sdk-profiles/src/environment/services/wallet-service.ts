import { pqueue } from "../../helpers/queue";
import { Profile } from "../../profiles/profile";
import { ProfileRepository } from "../../repositories/profile-repository";
import { container } from "../container";
import { Identifiers } from "../container.models";

export class WalletService {
	public async syncAll(): Promise<void> {
		const promises: (() => Promise<void>)[] = [];

		for (const profile of container.get<ProfileRepository>(Identifiers.ProfileRepository).values()) {
			for (const wallet of profile.wallets().values()) {
				promises.push(() => wallet?.syncIdentity());
				promises.push(() => wallet?.syncVotes());
			}
		}

		await pqueue(promises);
	}

	public async syncByProfile(profile: Profile): Promise<void> {
		const promises: (() => Promise<void>)[] = [];

		for (const wallet of profile.wallets().values()) {
			promises.push(() => wallet?.syncIdentity());
			promises.push(() => wallet?.syncVotes());
		}

		await pqueue(promises);
	}
}
