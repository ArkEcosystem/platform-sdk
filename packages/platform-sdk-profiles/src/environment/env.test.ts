import "jest-extended";

import { Coins } from "@arkecosystem/platform-sdk";
import { ARK } from "@arkecosystem/platform-sdk-ark";
import { BTC } from "@arkecosystem/platform-sdk-btc";
import { ETH } from "@arkecosystem/platform-sdk-eth";
import { Request } from "@arkecosystem/platform-sdk-http-got";
import { removeSync } from "fs-extra";
import nock from "nock";
import { resolve } from "path";

import storageData from "../../test/fixtures/env-storage.json";
import corruptedStorageData from "../../test/fixtures/env-storage-corrupted.json";
import { identity } from "../../test/fixtures/identity";
import { StubStorage } from "../../test/stubs/storage";
import { Profile } from "../profiles/profile";
import { DataRepository } from "../repositories/data-repository";
import { ProfileRepository } from "../repositories/profile-repository";
import { container } from "./container";
import { Identifiers } from "./container.models";
import { Environment } from "./env";
import { WalletService } from "./services/wallet-service";
import { MemoryStorage } from "./storage/memory";

let subject: Environment;

beforeAll(() => {
	nock.disableNetConnect();

	nock(/.+/)
		.get("/api/node/configuration")
		.reply(200, require("../../test/fixtures/client/configuration.json"))
		.get("/api/node/configuration/crypto")
		.reply(200, require("../../test/fixtures/client/cryptoConfiguration.json"))
		.get("/api/node/syncing")
		.reply(200, require("../../test/fixtures/client/syncing.json"))
		.get("/api/peers")
		.reply(200, require("../../test/fixtures/client/peers.json"))
		.get("/api/node/fees")
		.query(true)
		.reply(200, require("../../test/fixtures/client/node-fees.json"))
		.get("/api/transactions/fees")
		.query(true)
		.reply(200, require("../../test/fixtures/client/transaction-fees.json"))
		.get("/api/delegates")
		.query(true)
		.reply(200, require("../../test/fixtures/client/delegates-2.json"))
		.get("/ArkEcosystem/common/master/devnet/known-wallets-extended.json")
		.reply(200, [
			{
				type: "team",
				name: "ACF Hot Wallet",
				address: "AagJoLEnpXYkxYdYkmdDSNMLjjBkLJ6T67",
			},
			{
				type: "team",
				name: "ACF Hot Wallet (old)",
				address: "AWkBFnqvCF4jhqPSdE2HBPJiwaf67tgfGR",
			},
			{
				type: "exchange",
				name: "Binance",
				address: "AFrPtEmzu6wdVpa2CnRDEKGQQMWgq8nE9V",
			},
		])
		.get("/api/wallets/D61mfSggzbvQgTUe6JhYKH2doHaqJ3Dyib")
		.reply(200, require("../../test/fixtures/client/wallet.json"))
		.get(/.+/)
		.query(true)
		.reply(200, (url) => {
			console.log("getting url", url);
			return { meta: {}, data: {} };
		})
		.persist();

	container.set(Identifiers.HttpClient, new Request());
	container.set(Identifiers.Coins, { ARK, BTC, ETH });
});

beforeEach(async () => {
	removeSync(resolve(__dirname, "../../test/stubs/env.json"));

	subject = new Environment({ coins: { ARK, BTC, ETH }, httpClient: new Request(), storage: new StubStorage() });
	await subject.verify();
	await subject.boot();
	await subject.persist();
});

it("should have a profile repository", async () => {
	expect(subject.profiles()).toBeInstanceOf(ProfileRepository);
});

it("should have a data repository", async () => {
	expect(subject.data()).toBeInstanceOf(DataRepository);
});

it("should have available networks", async () => {
	const coins: Record<string, Coins.CoinSpec> = { ARK, BTC, ETH };

	for (const network of subject.availableNetworks()) {
		expect(network.toObject()).toEqual(coins[network.coin()].manifest.networks[network.id()]);
	}
});

it("should create a profile with data and persist it when instructed to do so", async () => {
	/**
	 * Save data in the current environment.
	 */

	const profile = subject.profiles().create("John Doe");

	// Create a Contact
	profile.contacts().create("Jane Doe");

	// Create a Wallet
	await profile.wallets().importByMnemonic(identity.mnemonic, "ARK", "ark.devnet");

	// Create a Notification
	profile.notifications().push({
		icon: "warning",
		name: "Ledger Update Available",
		body: "...",
		type: "ledger",
		action: "Read Changelog",
	});

	// Create a DataEntry
	profile.data().set("key", "value");

	// Create a Setting
	profile.settings().set("ADVANCED_MODE", "value");

	// Create a Global DataEntry
	subject.data().set("key", "value");

	// Persist the data for the next instance to use.
	await subject.persist();

	/**
	 * Load data that the previous environment instance saved.
	 */

	const newEnv = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await newEnv.verify();
	await newEnv.boot();

	const newProfile = newEnv.profiles().findById(profile.id());

	expect(newProfile).toBeInstanceOf(Profile);

	await newProfile.restore();

	expect(newProfile.wallets().keys()).toHaveLength(1);
	expect(newProfile.contacts().keys()).toHaveLength(1);
	expect(newProfile.notifications().keys()).toHaveLength(1);
	expect(newProfile.data().all()).toEqual({ key: "value" });
	expect(newProfile.settings().all()).toEqual({
		ADVANCED_MODE: "value",
		AUTOMATIC_SIGN_OUT_PERIOD: 15,
		BIP39_LOCALE: "english",
		EXCHANGE_CURRENCY: "BTC",
		LEDGER_UPDATE_METHOD: false,
		LOCALE: "en-US",
		MARKET_PROVIDER: "cryptocompare",
		NAME: "John Doe",
		SCREENSHOT_PROTECTION: true,
		THEME: "light",
		TIME_FORMAT: "h:mm A",
		USE_CUSTOM_PEER: false,
		USE_MULTI_PEER_BROADCAST: false,
		USE_TEST_NETWORKS: false,
	});
});

it("should boot the environment from fixed data", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await env.verify(storageData);
	await env.boot();

	const newProfile = env.profiles().findById("b999d134-7a24-481e-a95d-bc47c543bfc9");

	await newProfile.restore();

	expect(newProfile).toBeInstanceOf(Profile);
	expect(newProfile.wallets().keys()).toHaveLength(1);
	expect(newProfile.contacts().keys()).toHaveLength(1);
	expect(newProfile.notifications().keys()).toHaveLength(1);
	expect(newProfile.data().all()).toEqual({ key: "value" });
	expect(newProfile.settings().all()).toEqual({ ADVANCED_MODE: "value", NAME: "John Doe" });

	const restoredWallet = newProfile.wallets().findById("ac38fe6d-4b67-4ef1-85be-17c5f6841129");
	expect(restoredWallet.settings().all()).toEqual({ ALIAS: "Johnathan Doe", AVATAR: "..." });
	expect(restoredWallet.alias()).toBe("Johnathan Doe");
});

it("should boot with empty storage data", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await env.verify({ profiles: storageData.profiles, data: {} });
	await env.boot();
});

it("should boot with empty storage profiles", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await env.verify({ profiles: {}, data: { key: "value" } });
	await env.boot();
});

it("should create preselected storage given storage option as string", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: "memory" });
	expect(container.get(Identifiers.Storage)).toBeInstanceOf(MemoryStorage);
});

it("should throw error when calling usedCoinsWithNetworks without verify first", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	expect(() => env.usedCoinsWithNetworks()).toThrowError("Please call [verify] before looking up profile data.");
});

it("should get a list of used coins and networks", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await env.verify(storageData);
	await env.boot();

	expect(env.usedCoinsWithNetworks()).toEqual({ ARK: ["ark.devnet"] });
});

it("should throw error when calling boot without verify first", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await expect(env.boot()).rejects.toThrowError("Please call [verify] before booting the environment.");
});

it("should get available coins", async () => {
	expect(subject.coins().values()).toEqual([]);
});

it("#exchangeRates", async () => {
	await subject.exchangeRates().syncAll();
	expect(subject.exchangeRates()).toEqual({});
});

it("#fees", async () => {
	await subject.fees().sync("ARK", "ark.devnet");
	expect(Object.keys(subject.fees().all("ARK", "ark.devnet"))).toHaveLength(14);
});

it("#delegates", async () => {
	await subject.delegates().sync("ARK", "ark.devnet");
	expect(subject.delegates().all("ARK", "ark.devnet")).toHaveLength(100);
});

it("#knownWallets", async () => {
	await subject.knownWallets().syncAll();
	expect(subject.knownWallets().is("ark.devnet", "unknownWallet")).toBeFalse();
});

it("#wallets", async () => {
	expect(subject.wallets()).toBeInstanceOf(WalletService);
});

it("#coin", async () => {
	await expect(subject.coin("ARK", "ark.devnet")).resolves.toBeInstanceOf(Coins.Coin);
});

it("#registerCoin", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await env.verify(storageData);
	await env.boot();

	env.registerCoin("BTC", BTC);
	expect(() => env.registerCoin("BTC", BTC)).toThrowError(/is already registered/);
});

it("should fail verification", async () => {
	const env = new Environment({ coins: { ARK }, httpClient: new Request(), storage: new StubStorage() });
	await expect(env.verify(corruptedStorageData)).rejects.toThrowError("Terminating due to corrupted state.");
});

it("#migrate", async () => {
	expect(() => subject.migrate(storageData, "2.0.0")).not.toThrow();
});
