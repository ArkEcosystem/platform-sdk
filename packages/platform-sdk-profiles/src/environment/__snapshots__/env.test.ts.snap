// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have available networks 1`] = `
Object {
  "coin": "ARK",
  "crypto": Object {
    "expirationType": "height",
    "signingMethods": Object {
      "mnemonic": true,
      "wif": true,
    },
    "slip44": 111,
  },
  "currency": Object {
    "symbol": "Ѧ",
    "ticker": "ARK",
  },
  "explorer": "https://explorer.ark.io/",
  "featureFlags": Object {
    "Client": Object {
      "broadcast": true,
      "configuration": true,
      "delegate": true,
      "delegates": true,
      "fees": true,
      "syncing": true,
      "transaction": true,
      "transactions": true,
      "voters": true,
      "votes": true,
      "wallet": true,
      "wallets": true,
    },
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Fee": Object {
      "all": true,
    },
    "Identity": Object {
      "address": Object {
        "mnemonic": true,
        "multiSignature": true,
        "privateKey": true,
        "publicKey": true,
        "wif": true,
      },
      "keyPair": Object {
        "mnemonic": true,
        "privateKey": false,
        "wif": true,
      },
      "privateKey": Object {
        "mnemonic": true,
        "wif": true,
      },
      "publicKey": Object {
        "mnemonic": true,
        "multiSignature": true,
        "wif": true,
      },
      "wif": Object {
        "mnemonic": true,
      },
    },
    "Internal": Object {
      "fastDelegateSync": true,
    },
    "Ledger": Object {
      "getPublicKey": true,
      "getVersion": true,
      "signMessage": true,
      "signTransaction": true,
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Miscellaneous": Object {
      "dynamicFees": true,
      "memo": true,
    },
    "Peer": Object {
      "search": true,
    },
    "Transaction": Object {
      "delegateRegistration": true,
      "delegateResignation": true,
      "htlcClaim": false,
      "htlcLock": false,
      "htlcRefund": false,
      "ipfs": true,
      "multiPayment": true,
      "multiSignature": true,
      "secondSignature": true,
      "transfer": true,
      "vote": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 51,
      "enabled": true,
      "maximumPerTransaction": 1,
      "maximumPerWallet": 1,
    },
  },
  "id": "ark.mainnet",
  "knownWallets": "https://raw.githubusercontent.com/ArkEcosystem/common/master/mainnet/known-wallets-extended.json",
  "name": "Mainnet",
  "networking": Object {
    "hosts": Array [
      "https://wallets.ark.io",
    ],
    "hostsMultiSignature": Array [
      "https://musig1.ark.io",
    ],
  },
  "transactionTypes": Array [
    "delegate-registration",
    "delegate-resignation",
    "htlc-claim",
    "htlc-lock",
    "htlc-refund",
    "ipfs",
    "multi-payment",
    "multi-signature",
    "second-signature",
    "transfer",
    "vote",
  ],
  "type": "live",
}
`;

exports[`should have available networks 2`] = `
Object {
  "coin": "ARK",
  "crypto": Object {
    "expirationType": "height",
    "signingMethods": Object {
      "mnemonic": true,
      "wif": true,
    },
    "slip44": 1,
  },
  "currency": Object {
    "symbol": "DѦ",
    "ticker": "DARK",
  },
  "explorer": "https://dexplorer.ark.io/",
  "featureFlags": Object {
    "Client": Object {
      "broadcast": true,
      "configuration": true,
      "delegate": true,
      "delegates": true,
      "fees": true,
      "syncing": true,
      "transaction": true,
      "transactions": true,
      "voters": true,
      "votes": true,
      "wallet": true,
      "wallets": true,
    },
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Fee": Object {
      "all": true,
    },
    "Identity": Object {
      "address": Object {
        "mnemonic": true,
        "multiSignature": true,
        "privateKey": true,
        "publicKey": true,
        "wif": true,
      },
      "keyPair": Object {
        "mnemonic": true,
        "privateKey": false,
        "wif": true,
      },
      "privateKey": Object {
        "mnemonic": true,
        "wif": true,
      },
      "publicKey": Object {
        "mnemonic": true,
        "multiSignature": true,
        "wif": true,
      },
      "wif": Object {
        "mnemonic": true,
      },
    },
    "Internal": Object {
      "fastDelegateSync": true,
    },
    "Ledger": Object {
      "getPublicKey": true,
      "getVersion": true,
      "signMessage": true,
      "signTransaction": true,
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Miscellaneous": Object {
      "dynamicFees": true,
      "memo": true,
    },
    "Peer": Object {
      "search": true,
    },
    "Transaction": Object {
      "delegateRegistration": true,
      "delegateResignation": true,
      "htlcClaim": true,
      "htlcLock": true,
      "htlcRefund": true,
      "ipfs": true,
      "multiPayment": true,
      "multiSignature": true,
      "secondSignature": true,
      "transfer": true,
      "vote": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 51,
      "enabled": true,
      "maximumPerTransaction": 1,
      "maximumPerWallet": 1,
    },
  },
  "id": "ark.devnet",
  "knownWallets": "https://raw.githubusercontent.com/ArkEcosystem/common/master/devnet/known-wallets-extended.json",
  "name": "Devnet",
  "networking": Object {
    "hosts": Array [
      "https://dwallets.ark.io",
    ],
    "hostsMultiSignature": Array [
      "https://dmusig1.ark.io",
    ],
  },
  "transactionTypes": Array [
    "delegate-registration",
    "delegate-resignation",
    "htlc-claim",
    "htlc-lock",
    "htlc-refund",
    "ipfs",
    "multi-payment",
    "multi-signature",
    "second-signature",
    "transfer",
    "vote",
  ],
  "type": "test",
}
`;

exports[`should have available networks 3`] = `
Object {
  "coin": "Compendia",
  "crypto": Object {
    "expirationType": "height",
    "signingMethods": Object {
      "mnemonic": true,
      "wif": true,
    },
    "slip44": 543,
  },
  "currency": Object {
    "symbol": "ß",
    "ticker": "BIND",
  },
  "explorer": "https://bindscan.io/",
  "featureFlags": Object {
    "Client": Object {
      "broadcast": true,
      "configuration": true,
      "delegate": true,
      "delegates": true,
      "fees": true,
      "syncing": true,
      "transaction": true,
      "transactions": true,
      "voters": true,
      "votes": true,
      "wallet": true,
      "wallets": true,
    },
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Fee": Object {
      "all": true,
    },
    "Identity": Object {
      "address": Object {
        "mnemonic": true,
        "multiSignature": true,
        "privateKey": true,
        "publicKey": true,
        "wif": true,
      },
      "keyPair": Object {
        "mnemonic": true,
        "privateKey": false,
        "wif": true,
      },
      "privateKey": Object {
        "mnemonic": true,
        "wif": true,
      },
      "publicKey": Object {
        "mnemonic": true,
        "multiSignature": true,
        "wif": true,
      },
      "wif": Object {
        "mnemonic": true,
      },
    },
    "Internal": Object {
      "fastDelegateSync": true,
    },
    "Ledger": Object {
      "getPublicKey": true,
      "getVersion": true,
      "signMessage": true,
      "signTransaction": true,
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Miscellaneous": Object {
      "dynamicFees": true,
      "memo": true,
    },
    "Peer": Object {
      "search": true,
    },
    "Transaction": Object {
      "delegateRegistration": true,
      "delegateResignation": true,
      "htlcClaim": false,
      "htlcLock": false,
      "htlcRefund": false,
      "ipfs": true,
      "multiPayment": true,
      "multiSignature": true,
      "secondSignature": true,
      "transfer": true,
      "vote": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 47,
      "enabled": true,
      "maximumPerTransaction": 1,
      "maximumPerWallet": 1,
    },
  },
  "id": "compendia.mainnet",
  "name": "Mainnet",
  "networking": Object {
    "hosts": Array [
      "https://apis.compendia.org",
    ],
  },
  "transactionTypes": Array [
    "delegate-registration",
    "delegate-resignation",
    "htlc-claim",
    "htlc-lock",
    "htlc-refund",
    "ipfs",
    "multi-payment",
    "multi-signature",
    "second-signature",
    "transfer",
    "vote",
  ],
  "type": "live",
}
`;

exports[`should have available networks 4`] = `
Object {
  "coin": "Bitcoin",
  "crypto": Object {
    "expirationType": "height",
    "signingMethods": Object {
      "mnemonic": true,
      "wif": true,
    },
    "slip44": 0,
  },
  "currency": Object {
    "symbol": "Ƀ",
    "ticker": "BTC",
  },
  "explorer": "https://blockstream.info/",
  "featureFlags": Object {
    "Client": Object {
      "broadcast": true,
      "transaction": true,
      "wallet": true,
    },
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
      "bip49": true,
      "bip84": true,
    },
    "Identity": Object {
      "address": Object {
        "mnemonic": true,
        "multiSignature": true,
        "privateKey": true,
        "publicKey": true,
        "wif": true,
      },
      "keyPair": Object {
        "mnemonic": true,
        "privateKey": true,
        "wif": true,
      },
      "privateKey": Object {
        "mnemonic": true,
        "wif": true,
      },
      "publicKey": Object {
        "mnemonic": true,
        "wif": true,
      },
      "wif": Object {
        "mnemonic": true,
      },
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Miscellaneous": Object {
      "utxo": true,
    },
    "Transaction": Object {
      "transfer": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 0,
      "enabled": false,
      "maximumPerTransaction": 0,
      "maximumPerWallet": 0,
    },
  },
  "id": "btc.livenet",
  "name": "Livenet",
  "networking": Object {
    "hosts": Array [
      "https://coins.com/api/btc",
    ],
  },
  "transactionTypes": Array [
    "transfer",
  ],
  "type": "live",
}
`;

exports[`should have available networks 5`] = `
Object {
  "coin": "Bitcoin",
  "crypto": Object {
    "expirationType": "height",
    "signingMethods": Object {
      "mnemonic": true,
      "wif": true,
    },
    "slip44": 0,
  },
  "currency": Object {
    "symbol": "Ƀ",
    "ticker": "BTC",
  },
  "explorer": "https://blockstream.info/testnet/",
  "featureFlags": Object {
    "Client": Object {
      "broadcast": true,
      "transaction": true,
      "wallet": true,
    },
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
      "bip49": true,
      "bip84": true,
    },
    "Identity": Object {
      "address": Object {
        "mnemonic": true,
        "multiSignature": true,
        "privateKey": true,
        "publicKey": true,
        "wif": true,
      },
      "keyPair": Object {
        "mnemonic": true,
        "privateKey": true,
        "wif": true,
      },
      "privateKey": Object {
        "mnemonic": true,
        "wif": true,
      },
      "publicKey": Object {
        "mnemonic": true,
        "wif": true,
      },
      "wif": Object {
        "mnemonic": true,
      },
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Miscellaneous": Object {
      "utxo": true,
    },
    "Transaction": Object {
      "transfer": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 0,
      "enabled": false,
      "maximumPerTransaction": 0,
      "maximumPerWallet": 0,
    },
  },
  "id": "btc.testnet",
  "name": "Testnet",
  "networking": Object {
    "hosts": Array [
      "https://coins.com/api/btc",
    ],
  },
  "transactionTypes": Array [
    "transfer",
  ],
  "type": "test",
}
`;

exports[`should have available networks 6`] = `
Object {
  "coin": "Ethereum",
  "crypto": Object {
    "expirationType": "height",
    "networkId": "1",
    "signingMethods": Object {
      "mnemonic": true,
      "privateKey": true,
    },
    "slip44": 60,
  },
  "currency": Object {
    "symbol": "Ξ",
    "ticker": "ETH",
  },
  "explorer": "https://etherscan.io/",
  "featureFlags": Object {
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Identity": Object {
      "address": Object {
        "privateKey": true,
        "publicKey": true,
      },
      "keyPair": Object {
        "privateKey": true,
      },
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Transaction": Object {
      "transfer": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 0,
      "enabled": false,
      "maximumPerTransaction": 0,
      "maximumPerWallet": 0,
    },
  },
  "id": "eth.mainnet",
  "name": "Mainnet",
  "networking": Object {
    "hosts": Array [
      "https://platform.ark.io/api/eth",
    ],
  },
  "transactionTypes": Array [
    "transfer",
  ],
  "type": "live",
}
`;

exports[`should have available networks 7`] = `
Object {
  "coin": "Ethereum",
  "crypto": Object {
    "expirationType": "height",
    "networkId": "2",
    "signingMethods": Object {
      "mnemonic": true,
      "privateKey": true,
    },
    "slip44": 60,
  },
  "currency": Object {
    "symbol": "Ξ",
    "ticker": "ETH",
  },
  "explorer": "https://kovan.etherscan.io/",
  "featureFlags": Object {
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Identity": Object {
      "address": Object {
        "privateKey": true,
        "publicKey": true,
      },
      "keyPair": Object {
        "privateKey": true,
      },
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Transaction": Object {
      "transfer": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 0,
      "enabled": false,
      "maximumPerTransaction": 0,
      "maximumPerWallet": 0,
    },
  },
  "id": "eth.kovan",
  "name": "Kovan",
  "networking": Object {
    "hosts": Array [
      "https://coins.com/api/eth",
    ],
  },
  "transactionTypes": Array [
    "transfer",
  ],
  "type": "test",
}
`;

exports[`should have available networks 8`] = `
Object {
  "coin": "Ethereum",
  "crypto": Object {
    "expirationType": "height",
    "networkId": "3",
    "signingMethods": Object {
      "mnemonic": true,
      "privateKey": true,
    },
    "slip44": 60,
  },
  "currency": Object {
    "symbol": "Ξ",
    "ticker": "ETH",
  },
  "explorer": "https://ropsten.etherscan.io/",
  "featureFlags": Object {
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Identity": Object {
      "address": Object {
        "privateKey": true,
        "publicKey": true,
      },
      "keyPair": Object {
        "privateKey": true,
      },
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Transaction": Object {
      "transfer": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 0,
      "enabled": false,
      "maximumPerTransaction": 0,
      "maximumPerWallet": 0,
    },
  },
  "id": "eth.ropsten",
  "name": "Ropsten",
  "networking": Object {
    "hosts": Array [
      "https://coins.com/api/eth",
    ],
  },
  "transactionTypes": Array [
    "transfer",
  ],
  "type": "test",
}
`;

exports[`should have available networks 9`] = `
Object {
  "coin": "Ethereum",
  "crypto": Object {
    "expirationType": "height",
    "networkId": "4",
    "signingMethods": Object {
      "mnemonic": true,
      "privateKey": true,
    },
    "slip44": 60,
  },
  "currency": Object {
    "symbol": "Ξ",
    "ticker": "ETH",
  },
  "explorer": "https://rinkeby.etherscan.io/",
  "featureFlags": Object {
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Identity": Object {
      "address": Object {
        "privateKey": true,
        "publicKey": true,
      },
      "keyPair": Object {
        "privateKey": true,
      },
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Transaction": Object {
      "transfer": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 0,
      "enabled": false,
      "maximumPerTransaction": 0,
      "maximumPerWallet": 0,
    },
  },
  "id": "eth.rinkeby",
  "name": "Rinkeby",
  "networking": Object {
    "hosts": Array [
      "https://coins.com/api/eth",
    ],
  },
  "transactionTypes": Array [
    "transfer",
  ],
  "type": "test",
}
`;

exports[`should have available networks 10`] = `
Object {
  "coin": "Ethereum",
  "crypto": Object {
    "expirationType": "height",
    "networkId": "5",
    "signingMethods": Object {
      "mnemonic": true,
      "privateKey": true,
    },
    "slip44": 60,
  },
  "currency": Object {
    "symbol": "Ξ",
    "ticker": "ETH",
  },
  "explorer": "https://goerli.etherscan.io/",
  "featureFlags": Object {
    "Derivation": Object {
      "bip39": true,
      "bip44": true,
    },
    "Identity": Object {
      "address": Object {
        "privateKey": true,
        "publicKey": true,
      },
      "keyPair": Object {
        "privateKey": true,
      },
    },
    "Link": Object {
      "block": true,
      "transaction": true,
      "wallet": true,
    },
    "Message": Object {
      "sign": true,
      "verify": true,
    },
    "Transaction": Object {
      "transfer": true,
    },
  },
  "governance": Object {
    "voting": Object {
      "delegateCount": 0,
      "enabled": false,
      "maximumPerTransaction": 0,
      "maximumPerWallet": 0,
    },
  },
  "id": "eth.goerli",
  "name": "Goerli",
  "networking": Object {
    "hosts": Array [
      "https://coins.com/api/eth",
    ],
  },
  "transactionTypes": Array [
    "transfer",
  ],
  "type": "test",
}
`;
